### DISCUSSION TITLE
The type hint of Guild.voice_client

### DISCUSSION THREAD

**Discussion by tnthung** [Rapptz/discord.py#7445] [Q&A]
In source code and doc, the return type of method `Guild.voice_client` is `VoiceProtocal`. But the actual return type is `VoiceClient`. All examples on the web I can find are treating `Guild.voice_client` as `VoiceClient`, but the IDE type hint keeps telling me it's `VoiceProtocal` and doesn't have the method `is_playing`. I'm so confused now. Is it been a typo in the doc or been forgotten between revisions?

**Comment by apple502j** [comment]
`VoiceClient` is a subclass of `VoiceProtocol`. If you use custom subclass of `VoiceProtocol` (`cls=`) when you execute `VoiceChannel.connect`, then it doesn't have the `is_playing` method.

**Reply by tnthung** [reply]
I'm using bear discord.py to construct the bot with no additional custom wrapper. Everything works fine with the example I copied from web. Just the type hint thingy makes me uncomfortable a bit.

**Comment by Rapptz** [comment]
`VoiceProtocol` is the base class of `VoiceClient`. Someone can customise the behaviour by specifying their own subclass within `VoiceChannel.connect`.

> but the IDE type hint keeps telling me it's VoiceProtocal and doesn't have the method is_playing.

Do you mean the actual type hint in the code or the one that PyCharm extracts from the docstring? The former shouldn't have a problem in v2. If you're still on v1 I recommend using the [discord.py stubs by BryanForbes](https://github.com/bryanforbes/discord.py-stubs).

**Reply by tnthung** [reply]
Both I think. I'm not using any kind of wrapper between discord.py and my bot, and this is the part of `voice_client` defined in the `discord.Guild`. The return type hint is `Optional[VoiceProtocol]`.
```py
class Guild(Hashable):
    ...
    @property
    def voice_client(self) -> Optional[VoiceProtocol]:
        """Optional[:class:`VoiceProtocol`]: Returns the :class:`VoiceProtocol` associated with this guild, if any."""
        return self._state._get_voice_client(self.id)
```
But the examples online are doing something like:
```py
async def on_message(self, msg: Message):
    ...
    if msg.content == "play":
        msg.guild.voice_client.play(source)
```
To me, it looks weird since `VoiceProtocal` has no `play` method.

**Reply by bijij** [reply]
There's no practical way to have `Guild.voice_client` not be a `VoiceProtocol` unfortunately.

If it's bothering you enough I'd suggest looking at `typing.cast`, using an `assert` statement to type narrow, of simply using a ` # type: ignore` comment.

**Reply by tnthung** [reply]
@bijij But the second code block works fine, and `type(msg.guild.voice_client)` is indeed equal to `VoiceClient`.
> There's no practical way to have Guild.voice_client not be a VoiceProtocol unfortunately.

That's why I can't figure out what's going on. Sorry for my stupidity.

**Reply by bijij** [reply]
Yeah it's just a side-effect of the limitations of python type hinting.

The type at runtime is `VoiceClient`, but the type-checker will, unless you specify otherwise think it's `VoiceProtocol`

**Reply by tnthung** [reply]
@bijij I just tried the `typing.cast` you told me, and it works like a charm. I'm so appreciated. Thanks!!
